api/service/project_service.go

```diff
@@ -1,49 +1,143 @@
+	mlpQueryTimeoutSeconds = 30
+)
+
+// ProjectsService provides a set of methods to interact with the MLP / Merlin APIs
 type ProjectsService interface {
-	List(ctx context.Context, projectName string) (mlp.Projects, error)
+	// List lists available projects, optionally filtered by given project `name`
+	List(ctx context.Context, name string) (mlp.Projects, error)
+	// GetByID gets the project matching the provided id.
 	GetByID(ctx context.Context, projectID int32) (mlp.Project, error)
-	GetByName(ctx context.Context, projectName string) (mlp.Project, error)
 }
 
-func NewProjectsService(mlpAPIClient mlp.APIClient) ProjectsService {
-	return &projectsService{
-		mlpAPIClient: mlpAPIClient,
+type projectsService struct {
+	mlpClient mlp.APIClient
+	cache     *freecache.Cache
+}
+
+// NewProjectsService returns a service that retrieves information that is shared across MLP projects.
+func NewProjectsService(mlpAPIClient mlp.APIClient) (ProjectsService, error) {
+	svc := &projectsService{
+		mlpClient: mlpAPIClient,
+		cache:     freecache.NewCache(projectCacheSizeBytes),
+	}
+
+	err := svc.refreshProjects(context.Background())
+	if err != nil {
+		return nil, err
 	}
+	return svc, nil
 }
 
-type projectsService struct {
-	mlpAPIClient mlp.APIClient
+func (ps projectsService) List(ctx context.Context, name string) (mlp.Projects, error) {
+	err := ps.refreshProjects(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return ps.listProjects(name)
 }
 
-func (service *projectsService) List(ctx context.Context, projectName string) (mlp.Projects, error) {
-	return service.mlpAPIClient.ListProjects(ctx, projectName)
+// GetByID gets the project matching the provided id.
+// This method will hit the cache first, and if not found, will call Merlin once to get
+// the updated list of projects and refresh the cache, then try to get the value again.
+// If still not found, will return a freecache NotFound error.
+func (ps projectsService) GetByID(ctx context.Context, projectID int32) (mlp.Project, error) {
+	project, err := ps.getProject(projectID)
+	if err != nil {
+		err = ps.refreshProjects(ctx)
+		if err != nil {
+			return mlp.Project{}, err
+		}
+		project, err = ps.getProject(projectID)
+		if err != nil {
+			return mlp.Project{}, err
+		}
+	}
+	return *project, nil
+}
+
+func (ps projectsService) listProjects(name string) (mlp.Projects, error) {
+	projects := mlp.Projects{}
+	// List all items in the cache
+	iter := ps.cache.NewIterator()
+	for entry := iter.Next(); entry != nil; {
+		key := string(entry.Key)
+		if strings.HasPrefix(key, projectCacheKeyPrefix) {
+			var project mlpClient.Project
+			if err := json.Unmarshal(entry.Value, &project); err != nil {
+				return mlp.Projects{}, fmt.Errorf("Malformed project info found in the cache for key %s: %w", key, err)
+			} else {
+				// If either the name filter is not set or the project's name matches
+				if name == "" || project.Name == name {
+					projects = append(projects, project)
+				}
+			}
```

Comment:
nit, I prefer fewer branching here and faster return when filtered project found:

```
if err := json.Unmarshal(entry.Value, &project); err != nil {
        return mlp.Projects{}, fmt.Errorf("Malformed project info found in the cache for key %s: %w", key, err)
}

if project.Name == name {
	return mlp.Projects{project}
}

projects = append(projects, project)
```